encryptor/queryDataEncryptor.go

```diff
@@ -313,6 +316,187 @@ func (encryptor *QueryDataEncryptor) OnQuery(query base.OnQueryObject) (base.OnQ
 	return query, false, nil
 }
 
+// OnBind process bound values for prepared statement based on TableSchemaStore.
+func (encryptor *QueryDataEncryptor) OnBind(statement sqlparser.Statement, values []base.BoundValue) ([]base.BoundValue, bool, error) {
+	newValues := values
+	changed := false
+	var err error
+	switch statement := statement.(type) {
+	case *sqlparser.Insert:
+		newValues, changed, err = encryptor.encryptInsertValues(statement, values)
+	case *sqlparser.Update:
+		newValues, changed, err = encryptor.encryptUpdateValues(statement, values)
+	}
+	if err != nil {
+		return values, false, err
+	}
+	return newValues, changed, nil
+}
+
+func (encryptor *QueryDataEncryptor) encryptInsertValues(insert *sqlparser.Insert, values []base.BoundValue) ([]base.BoundValue, bool, error) {
+	logrus.Debugln("QueryDataEncryptor.encryptInsertValues")
+	tableName := insert.Table.Name
+	// Look for the schema of the table where the INSERT happens.
+	// If we don't have a schema then we don't know what to encrypt, so do nothing.
+	schema := encryptor.schemaStore.GetTableSchema(tableName.String())
+	if schema == nil {
+		logrus.WithField("table", tableName).Debugln("No encryption schema")
+		return values, false, nil
+	}
+
+	// Gather column names from the INSERT query. If there are no columns in the query,
+	// expect a complete list of colums to be available in the schema.
+	var columns []string
+	if len(insert.Columns) > 0 {
+		columns = make([]string, len(insert.Columns))
+		for i, column := range insert.Columns {
+			columns[i] = column.String()
+		}
+	} else if cols := schema.Columns(); len(cols) > 0 {
+		columns = cols
+	}
+
+	changed := false
+	newValues := make([]base.BoundValue, len(values))
+	copy(newValues, values)
+
+	// If there is no column schema available, we can't encrypt values.
+	if len(columns) > 0 {
+		// We can also only process simple queries of the form
+		//
+		//     INSERT INTO table(column...) VALUES ($1, $2, 'static value'...);
+		//
+		// That is, where placeholders uniquely identify the column and used directly
+		// as inserted values of a single row. We don't support functions, casts,
+		// inserting multiple values and query results, etc.
+		//
+		// Walk through the query to find out which placeholders stand for which columns.
+		placeholders := make(map[int]string)
+		switch rows := insert.Rows.(type) {
+		case sqlparser.Values:
+			if len(rows) == 1 {
```

Comment:
why we can't handle case with several values? queries like `insert into table(...) values (), (), ();`?

Reply:
> why we can't handle case with several values?

Uh... Now that you ask this... There does not seem to be a reason why not. The only thing that we care about is the placeholders being uniquely mapped onto columns. Which the code does not seems to check, btw. I'll fix this, and will add support for multiple value clauses.

Reply:
I've extracted placeholder processing into a utility method shared between INSERT and UPDATE statements. I've also added additional safeguards there which should allow multiple VALUES in INSERT queries. Provided that those clauses are consistent. That is, we won't be able to handle
```
INSERT ... VALUES ($1, $2),  ($2, $1)
```
where we can't figure out what columns the values are for. However, something like
```
INSERT ... VALUES ($1, $2),  ($3, $4),  ($5, $6)
```
or
```
INSERT ... VALUES ($1, $2, 'foo'),  ($1, $2, 'bar')
```
should work.